hw/ip/otbn/dv/otbnsim/sim/state.py

```diff
@@ -257,51 +186,86 @@ def commit(self) -> None:
         if self.stalled:
             return
 
-        self.intreg.commit()
-        self.pc.set(self.pc_next
-                    if self.pc_next is not None
-                    else self.pc.value + 4)
+        self.gprs.commit()
+        self.pc = self.pc_next if self.pc_next is not None else self.pc + 4
         self.pc_next = None
         self.dmem.commit()
         self.loop_stack.commit()
         self.ext_regs.commit()
         self.wsrs.commit()
         self.csrs.commit()
-        self.wreg.commit()
+        self.wdrs.commit()
 
     def start(self) -> None:
         '''Set the running flag and the ext_reg busy flag'''
         self.ext_regs.set_bits('STATUS', 1 << 0)
         self.running = True
 
-    def get_wr_quarterword(self, wridx: int, qwsel: int) -> int:
-        assert 0 <= wridx <= 31
+    def get_quarter_word_unsigned(self, idx: int, qwsel: int) -> int:
+        '''Select a 64-bit quarter of a wide register.
+
+        The bits are interpreted as an unsigned value.
+
+        '''
+        assert 0 <= idx <= 31
         assert 0 <= qwsel <= 3
-        mask = (1 << 64) - 1
-        return (int(self.wreg[wridx]) >> (qwsel * 64)) & mask
+        full_val = self.wdrs.get_reg(idx).read_unsigned()
+        return (full_val >> (qwsel * 64)) & ((1 << 64) - 1)
+
+    def set_half_word_unsigned(self, idx: int, hwsel: int, value: int) -> None:
+        '''Set the low or high 128-bit half of a wide register to value.
+
+        The value should be unsigned.
 
-    def set_wr_halfword(self, wridx: int, value: int, hwsel: int) -> None:
-        assert 0 <= wridx <= 31
-        assert (value >> 128) == 0
+        '''
+        assert 0 <= idx <= 31
         assert 0 <= hwsel <= 1
+        assert 0 <= value <= (1 << 128) - 1
+
+        shift = 128 * hwsel
+        shifted_input = value << shift
+        mask = ((1 << 128) - 1) << shift
 
-        mask = ((1 << 128) - 1) << (0 if hwsel else 128)
-        curr = int(self.wreg[wridx]) & mask
-        valpos = value << 128 if hwsel else value
-        self.wreg[wridx] = curr | valpos
+        old_val = self.wdrs.get_reg(idx).read_unsigned()
+        new_val = (old_val & ~mask) | shifted_input
+        self.wdrs.get_reg(idx).write_unsigned(new_val)
 
     @staticmethod
     def add_with_carry(a: int, b: int, carry_in: int) -> Tuple[int, FlagReg]:
```

Comment:
Wouldn't it make more sense now to also use the `Reg` type here (arguments of type `Reg`)?
We only use this in the context of instructions execution and there after reading `wdr`s, right?

Reply:
The `b` argument to `add_with_carry` is usually shifted. See, for example, the call site at `BNADD.execute`:
```python
        a = state.wdrs.get_reg(self.wrs1).read_unsigned()
        b = state.wdrs.get_reg(self.wrs2).read_unsigned()
        b_shifted = logical_byte_shift(b, self.shift_type, self.shift_bytes)

        (result, flags) = state.add_with_carry(a, b_shifted, 0)
```
I don't like the behaviour that the previous code had, where you'd construct `Register` types for intermediate values, so much prefer `logical_byte_shift` to return an integer.